src/System.Net.Sockets/src/System/Net/Sockets/Socket.cs

```diff
@@ -6128,8 +6194,23 @@ private static object PostOneBeginConnect(MultipleAddressConnectAsyncResult cont
                 // MSRC 11081 - Do the necessary security demand
                 context.socket.CheckCacheRemote(ref endPoint, true);
 
-                IAsyncResult connectResult = context.socket.UnsafeBeginConnect(endPoint,
-                    new AsyncCallback(MultipleAddressConnectCallback), context);
+                var asyncCallback = new AsyncCallback(MultipleAddressConnectCallback);
+
+                IAsyncResult connectResult;
+                if (SocketPal.SupportsMultipleConnectAttempts || context.isUserConnectAttempt)
+                {
+                    connectResult = context.socket.UnsafeBeginConnect(endPoint, asyncCallback, context);
+                }
+                else
+                {
+                    context.lastAttemptSocket = new Socket(context.socket._addressFamily, context.socket._socketType, context.socket._protocolType);
+                    if (context.socket.IsDualMode)
+                    {
+                        context.lastAttemptSocket.DualMode = true;
+                    }
+
+                    connectResult = context.lastAttemptSocket.UnsafeBeginConnect(endPoint, asyncCallback, context);
```

Comment:
Could you refactor this slightly to avoid needing the same basic line in both branches, e.g.
```C#
IAsyncResult connectResult;
Socket connectSocket = context.socket;
if (!SocketPal.SupportsMultipleConnectAttempts && context.IsConnectAttempt)
{
    context.lastAttemptSocket = connectSocket = new Socket(context.socket._addressFamily, context.socket._socketType, context.socket._protocolType) { DualMode = context.socket.IsDualMode };
}
connectSocket.UnsafeBeginConnect(endPoint, asyncCallback, context);
```
or something like that?